[GEN][ZH] Fix various memory leaks

Mauller opened the pull request:
**Merge with Rebase**

Fix freeing some reference counted pointers so that underlying memory is properly freed.
Also fix the cleanup in some classes that were not clearing dependent objects.

~~Will be checking for more so left this as draft for the moment.~~
EDIT: This should be good to merge now, the memory leaks it fixes are some of the most prominent that i came across.

Fixes:

- Gen + ZH - ~W3DLaserDraw() was leaking materials.
- Gen + ZH - ~ProductionUpdate() was leaking production queue items if game exited before buildings sold or destroyed.
- ZH - Snow system was leaking on shutdown, wouldn't affect runtime but means a clean shutdown.
- Gen + ZH - ~W3DTankDraw() was leaking particle systems, therefore leaking particles as well.
- Gen + ZH - ~MeshModelClass() had an extra call to free directx8 references in the wrong place, this could cause other dependent assets to not be deallocated properly within reset(). And Generals behaviour was freeing data in the wrong order.
- Gen + ZH - Win32LocalFileSystem was leaking Win32File handlers due to creating a handler and returning early on a null filename before deleting it. Moved the creation of the file handler after the early return and where it is needed / used.
- Gen - ~MeshModelClass() was calling the wrong delete method when deallocating GapFiller data. Fixed already in ZH.
- Gen + ZH - Win32BigFileSystem was leaking Big32File handlers due to creating a handler and returning early on a null filename before deleting it. Moved the creation of the file handler after the early return and where it is needed / used.
- Gen + ZH - ~W3DDisplay() was not releasing its debug display string so it could prevent DisplayStringManager from being cleanly destroyed.

Mauller commented:
Small update, found that production items in a production queue could leak if there were production queues in progress and a game was exited or finished before the queue finished or the building was sold or destroyed.

Mauller commented:
Small update where some references could be removed too early and some W3D asset manager glue was missing on mech models.

xezon commented on Generals/Code/Libraries/Source/WWVegas/WW3D2/meshmdl.cpp:
```diff
@@ -102,8 +102,6 @@ MeshModelClass::MeshModelClass(const MeshModelClass & that) :
 MeshModelClass::~MeshModelClass(void)
 {
-//	WWDEBUG_SAY(("Note: Mesh %s was never used\n",Get_Name()));
-	TheDX8MeshRenderer.Unregister_Mesh_Type(this);
```
Does not look like a memory leak fix.

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/W3DDevice/GameClient/Drawable/Draw/W3DTankDraw.cpp:
```diff
@@ -163,6 +163,10 @@ void W3DTankDraw::createEmitters( void )
 //-------------------------------------------------------------------------------------------------
 W3DTankDraw::~W3DTankDraw()
 {
+
```
can remove blank line.

Mauller commented:
Removed the MeshGeometryClass change as it was not necessary in the end.
Removed the W3DWaypoints change as i think the main thing i was originaly fixing in generals variant was fixed with a backport by Tomsons26.

The rest should be good to go now.

tomsons26 commented:
DisplayString instances may be leaking all over
neither m_benchmarkDisplayString or m_nativeDebugDisplay are even nulled in dtor.. 
m_displayStrings is just nulled but isn't deallocated..

Mauller commented:
Rebased with Main after the recent inclusions. Just need to look at what Tomson26 found today.

Mauller commented:
Rebased with recent Main.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/Snow.cpp:
```diff
@@ -99,6 +99,10 @@ SnowManager::~SnowManager()
 	m_startingHeights=NULL;
+
+	// TheSuperHackers Mauller 13/04/2025 Delete the instance of the weather settings.
+	((WeatherSetting*)TheWeatherSetting.getNonOverloadedPointer())->deleteInstance();
```
Is this safe? What if that pointer is null?

Mauller replied:
in theory it should always be there as it is created by the snow manager at startup. will put a conditional check on it jsut to be safe.

xezon commented on Generals/Code/GameEngineDevice/Source/W3DDevice/GameClient/W3DDisplay.cpp:
```diff
@@ -376,11 +376,18 @@ W3DDisplay::~W3DDisplay()
 	for (int i = 0; i < DisplayStringCount; i++)
 		TheDisplayStringManager->freeDisplayString(m_displayStrings[i]);
 
+	// TheSuperHackers @fix Mauller/Tomsons26 28/04/2025 Delete benchmark display string to allow the clean destruction of the DisplayStringManager
```
This text can be improved/simplified

Mauller replied:
done

xezon commented on Generals/Code/GameEngineDevice/Source/Win32Device/Common/Win32LocalFileSystem.cpp:
```diff
@@ -69,6 +68,9 @@ File * Win32LocalFileSystem::openFile(const Char *filename, Int access /* = 0 */
 	}
 
+	// TheSuperHackers Mauller 21/04/2025 Only create the new file instance when it is necassary or we leak file handles!
+	Win32LocalFile *file = newInstance( Win32LocalFile );	
```
This commit can be combined with the other commit that does the same kind of fix with ArchiveFile.

xezon commented on Generals/Code/GameEngineDevice/Source/Win32Device/Common/Win32LocalFileSystem.cpp:
```diff
@@ -69,6 +68,9 @@ File * Win32LocalFileSystem::openFile(const Char *filename, Int access /* = 0 */
 		}
 	}
 
+	// TheSuperHackers Mauller 21/04/2025 Only create the new file instance when it is necassary or we leak file handles!
```
This text can be improved/simplified.

Mauller replied:
Done

xezon commented on Generals/Code/GameEngineDevice/Source/W3DDevice/GameClient/Drawable/Draw/W3DLaserDraw.cpp:
```diff
@@ -234,6 +234,7 @@ W3DLaserDraw::~W3DLaserDraw( void )
 	}  // end for i
 
 	delete [] m_line3D;
+	REF_PTR_RELEASE(m_texture);
```
All other memory leak fixes have comments, so might as well add one here for consistency.

xezon reviewed (commented):
The commit titles and comments need some work. The way things are written is not always consistent, too complicated, not concise. Commit titles do not end on a dot. Comments can end on a dot (because they can have multiple sentences)

Mauller commented:
Updated based on reccomendations.

These commits and their comments were rather old now and before i started getting more of a hand of it. :P

xezon commented on Generals/Code/GameEngineDevice/Source/Win32Device/Common/Win32BIGFileSystem.cpp:
```diff
@@ -120,6 +118,8 @@ ArchiveFile * Win32BIGFileSystem::openArchiveFile(const Char *filename) {
 	fp->seek(0x10, File::START);
 	// read in each directory listing.
 	ArchivedFileInfo *fileInfo = NEW ArchivedFileInfo;
+	// TheSuperHackers @fix Mauller 23/04/2025 Create new file handler when necassary
```
type error

"Now creates file handler only when necessary to prevent leak and waste"

xezon replied:
Is this meant to "file handler". What does that mean?

Mauller replied:
A file handler is some object that manages access to a file for reading/writing/managment etc.

Mauller replied:
Renamed it to say "File Handle".

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/Snow.cpp:
This commit title has a superfluous dot.

How about use "memory leak" instead of mixed terminology "not being freed", "leaking", "cleanup", ...

For example

[GEN][ZH] Fix material memory leak in ~W3DLaserDraw() (#623)

Reason being, most noob visitors will understand what "memory leak" means but will potentially be oblivious to the terms "freed", "delete", "cleanup".

Mauller replied:
Thats fair, will take a look at rewording the commits again tomorrow.

xezon reviewed (commented):
Some more notes.

xezon commented on Generals/Code/GameEngineDevice/Source/Win32Device/Common/Win32BIGFileSystem.cpp:
```diff
@@ -120,6 +118,8 @@ ArchiveFile * Win32BIGFileSystem::openArchiveFile(const Char *filename) {
 	fp->seek(0x10, File::START);
 	// read in each directory listing.
 	ArchivedFileInfo *fileInfo = NEW ArchivedFileInfo;
+	// TheSuperHackers @fix Mauller 23/04/2025 Create new file handle when necassary to prevent memory leak
```
It still says "necassary"

Mauller replied:
Ah sod, that was me rushing to tweak it before heading to work.

xezon replied:
It was already there yesterday and this is what I referred to when I mentioned type error.

Mauller replied:
Corrected, i also thought you meant something else when saying "type error" as in data type instead of a "Typo" typography error / spelling mistake etc.

Mauller commented:
Should be good now, i will break out any others that i find into their own PR's in future.
